master/internal/rm/kubernetesrm/jobs.go

```diff
@@ -1929,6 +1937,103 @@ func (j *jobsService) removeEmptyNamespace(namespaceName string, clusterName str
 	return nil
 }
 
+func (j *jobsService) setResourceQuota(quota int, namespace string) error {
+	var k8sDeterminedLabel = map[string]string{determinedLabel: namespace}
+
+	k8sNamespace, err := j.clientSet.CoreV1().Namespaces().Get(context.TODO(), namespace,
+		metaV1.GetOptions{
+			TypeMeta: metaV1.TypeMeta{
+				Kind:       "Namespace",
+				APIVersion: "v1",
+			},
+		})
+	if err != nil {
+		return errors.Wrapf(err, "error finding namespace %s", namespace)
+	}
+
+	if _, ok := k8sNamespace.Labels[determinedLabel]; !ok {
+		return errors.Wrapf(err, "Cannot set quota on namespace %s. Namespace needs determined label",
+			namespace)
+	}
+
+	quotaName := namespace + "-quota"
+
+	// We want to patch the smallest quota if it is a determiend quota regardless of whether it's
+	// larger or smaller. If the quota does not correspond to the auto-generated namespace and is
+	// less than the current quota (the first non-Determined quota, we error out saying that they
+	// should remove that quota from Kubernetes and then try to set a Determined quota.
+	k8sQuotas, err := j.clientSet.CoreV1().ResourceQuotas(namespace).List(context.TODO(),
+		metaV1.ListOptions{})
+	quotas := k8sQuotas.Items
+	currentQuota := float64(quota)
+	var detQuota *k8sV1.ResourceQuota
+	for _, q := range quotas {
+		qResources := q.Spec.Hard
+		for name, quantity := range qResources {
+			if name == "requests."+ResourceTypeNvidia {
+				tmpQuota := quantity.AsApproximateFloat64()
+				q.Spec.Hard[name] = *resource.NewQuantity(int64(quota), resource.DecimalSI)
+				if q.Name == quotaName {
+					detQuota = &q
+				} else if tmpQuota < currentQuota {
+					return fmt.Errorf("cannot set quota %d on namespace %s, because this "+
+						"namespace consists of a Kubernetes quota with request limit %d that does not "+
+						"correspond to the auto-created namespace. Please remove this quota in "+
+						"Kubernetes before trying to raise the GPU request limit on the namespace.",
+						quota, namespace, int(tmpQuota))
+				}
+			}
+		}
+	}
+
+	if detQuota != nil {
+		// if currentQuota < float64(quota) {
+		// 	return fmt.Errorf("Cannot set quota because there already exists a quota in namespace %s of limit %d", namespaceName, currentQuota)
+		// }
+		detQuotaToByteArray, err := json.Marshal(detQuota)
+		if err != nil {
+			return errors.Wrapf(err, "error marshaling quota %s", detQuota.Name)
+		}
+		_, err = j.clientSet.CoreV1().ResourceQuotas(namespace).Patch(context.TODO(),
+			quotaName,
+			types.MergePatchType,
+			detQuotaToByteArray,
+			metaV1.PatchOptions{})
+
+		if err != nil {
+			return errors.Wrapf(err, "error applying patch to resource quota %s", quotaName)
+		}
+	} else {
```

Comment:
is this else necessary? looking at the flow of the func there's not any executable code outside of the return that's not in this else condition

Reply:
Yea! We need the else to create the resource quota if it doesn't exist already (rather than patch it, which is what occurs in the if-block.